swarm/bmt/bmt.go

```diff
@@ -244,130 +257,148 @@ func (t *Tree) Draw(hash []byte, d int) string {
 	return strings.Join(rows, "\n") + "\n"
 }
 
-// NewTree initialises the Tree by building up the nodes of a BMT
-// segment size is stipulated to be the size of the hash
-// segmentCount needs to be positive integer and does not need to be
-// a power of two and can even be an odd number
-// segmentSize * segmentCount determines the maximum chunk size
-// hashed using the tree
-func NewTree(hasher BaseHasher, segmentSize, segmentCount int) *Tree {
-	n := NewNode(0, 0, nil)
-	n.root = true
-	prevlevel := []*Node{n}
-	// iterate over levels and creates 2^level nodes
-	level := 1
+// newTree initialises a tree by building up the nodes of a BMT
+// - segment size is stipulated to be the size of the hash
+func newTree(segmentSize, depth int) *tree {
+	n := newNode(0, nil)
+	prevlevel := []*node{n}
+	// iterate over levels and creates 2^(depth-level) nodes
 	count := 2
-	for d := 1; d <= depth(segmentCount); d++ {
-		nodes := make([]*Node, count)
-		for i := 0; i < len(nodes); i++ {
+	for level := depth - 2; level >= 0; level-- {
+		nodes := make([]*node, count)
+		for i := 0; i < count; i++ {
 			parent := prevlevel[i/2]
-			t := NewNode(level, i, parent)
-			nodes[i] = t
+			nodes[i] = newNode(i, parent)
 		}
 		prevlevel = nodes
-		level++
 		count *= 2
 	}
-	// the datanode level is the nodes on the last level where
-	return &Tree{
-		leaves: prevlevel,
+	// the datanode level is the nodes on the last level
+	return &tree{
+		leaves:  prevlevel,
+		result:  make(chan []byte, 1),
+		segment: make([]byte, segmentSize),
+		section: make([]byte, 2*segmentSize),
 	}
 }
 
 // methods needed by hash.Hash
 
 // Size returns the size
 func (h *Hasher) Size() int {
-	return h.size
+	return h.pool.SegmentSize
 }
 
 // BlockSize returns the block size
 func (h *Hasher) BlockSize() int {
-	return h.blocksize
+	return h.pool.SegmentSize
+}
+
+// Hash hashes the data and the span using the bmt hasher
+func Hash(h *Hasher, span, data []byte) []byte {
```

Comment:
Shouldn't this be a method with pointer receiver on `Hasher`, rather than taking `h` as a first argument?

Reply:
no because `Hasher` will have a `Hash(int)` method. this function should just be a util in test and unexported.

Reply:
OK, but it is still exported right now?